### PR TITLE
[Sparse] Sparse matrix reduction C++ part

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Implemented `.sum()`, `.smin()`, `.smax()`, `.smean()` and `.prod()`.

First part of the work item #4865.  I'm putting the Python changes to another PR (#5013 ).

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: dgl_sparse/src/reduction.cc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  } else {
+    TORCH_CHECK(false, "unknown reduce function ", reduce);
+    return torch::Tensor();
+  }
+
+  // Create the output tensor with shape
+  //
+  //   [A.num_rows if dim == 1 else A.num_cols] + A.val.shape[1:]
+  //
+  // (BarclayII) I'm not sure how to get the Tensor's shape as a vector...
```
Comment (czkkkkkk) [inline dgl_sparse/src/reduction.cc:49]: Simply use `value.sizes().vec()`

File: dgl_sparse/src/reduction.cc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  std::vector<int64_t> output_shape(value_shape.begin(), value_shape.end());
+  std::vector<int64_t> view_dims(value_shape.size(), 1);
+  view_dims[0] = -1;
+  torch::Tensor idx;
+  if (dim == 0) {
+    output_shape[0] = coo->num_cols;
+    idx = coo->col.view(view_dims).expand_as(value);
+  } else if (dim == 1) {
+    output_shape[0] = coo->num_rows;
+    idx = coo->row.view(view_dims).expand_as(value);
```
Comment (czkkkkkk) [inline dgl_sparse/src/reduction.cc:60]: In the Python side, we do not need to expand `idx` to the shape of `value`. Maybe it is also not necessary in the C++ side? I am not very sure.

**Review by czkkkkkk — APPROVED** [review]

File: dgl_sparse/include/sparse/reduction.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+inline torch::Tensor ReduceMean(
+    const c10::intrusive_ptr<SparseMatrix>& A,
+    const torch::optional<int64_t>& dim = torch::nullopt) {
+  return Reduce(A, "smean", dim);
+}
+
+inline torch::Tensor ReduceProd(
+    const c10::intrusive_ptr<SparseMatrix>& A,
+    const torch::optional<int64_t>& dim = torch::nullopt) {
+  return Reduce(A, "prod", dim);
```
Comment (frozenbugs) [inline dgl_sparse/include/sparse/reduction.h:58]: sprod

File: dgl_sparse/include/sparse/reduction.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+namespace sparse {
+
+/**
+ * @brief Reduces a sparse matrix along the specified sparse dimension.
+ *
+ * @param A The sparse matrix.
+ * @param dim The sparse dimension to reduce along.  Must be either 0 (rows) or
+ * 1 (columns).
+ * @param reduce The reduce operator.  Must be either "sum", "smin", "smax",
+ * "mean", or "prod".
```
Comment (frozenbugs) [inline dgl_sparse/include/sparse/reduction.h:23]: smean / sprod

**Pull request merged**